feat(editor): Add workflow evaluation edit and list views (no-changelog)

OlegIvaniv opened the pull request:
This PR introduces a new Workflow Evaluation feature that allows users to create, list, and edit test definitions for workflows. The feature is currently behind the `025_workflow_evaluation` feature flag as it's still in development.

## Key Features

- New Evaluation tab in workflow view (visible when feature flag is enabled)
- Test definition management:
  - List view showing all tests with mocked metrics and status 
  - Create new tests with name, description, and metrics configuration
    - Metrics and description not stored by BE yet
  - Edit existing tests including name, tags, and evaluation workflow selection

## Implementation Details

- Added new API front-end client 
- Created new components for test management:
  - EvaluationListView: Main view for listing and managing tests
  - EvaluationEditView: Form for creating/editing test definitions
  - Various form components (MetricsInput, TagsInput, etc.)
- Integrated with existing tag management store
- Added new evaluations store for state management
- Implemented auto-save functionality for test edits

## Note

- Feature is disabled by default, enable with: `window.featureFlags.override('025_workflow_evaluation', true)`
- New routes added for test management:
  - `/workflow/:name/evaluation`
  - `/workflow/:name/evaluation/new`
  - `/workflow/:name/evaluation/:testId`


https://github.com/user-attachments/assets/cf5c584f-d492-4405-b386-7797cf325f5e


## Related Linear tickets, Github issues, and Community forum posts

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->


## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

OlegIvaniv committed:
Setup routes

OlegIvaniv committed:
Finish edit UI

OlegIvaniv committed:
Add test definitions API and store

- Created evaluations API endpoints
- Implemented evaluations store
- Updated NewWorkflowEvaluationView
- RenamedRun Test to Save Test

OlegIvaniv committed:
Refactor maxWidth to CSS variable

- Update GlobalExecutionsList.vue
- Update PageViewLayout.vue
- Define --content-container-width
- Update ProjectSettings.vue
- Update TemplatesView.vue

burivuhster committed:
wip: DB migration

OlegIvaniv committed:
Refactor workflow evaluation components and improve code organization

• Move components to dedicated files
• Create new EvaluationEditView component
• Update router configuration
• Relocate useEvaluationForm composable
• Add new types for evaluations

OlegIvaniv committed:
Refactor workflow evaluation list view and add new components

• Split EmptyState into separate component
• Create TestItem and TestsList components
• Update EvaluationListView with new components
• Add WORKFLOW_EVALUATION_EDIT to constants
• Improve styles and layout of test items

OlegIvaniv committed:
Implement feature flag for workflow evaluation experiment

- Add feature flag check
- Update UI to show/hide evaluation tab
- Fetch tags for test definitions
- Handle routing when feature is disabled
- Cleanup console logs

OlegIvaniv committed:
Refine UI for TestItem and TestsList components in WorkflowEvaluation

• Adjust padding and cursor for TestItem
• Update button sizes and spacing
• Modify TestItem title styling
• Remove size and type from TestsList button

OlegIvaniv committed:
Make TagsDropdown component more flexible and reusable

• Make createTag prop optional
• Add eventBus prop to TagsDropdown
• Update TagsInput to use TagsDropdown
• Remove unnecessary class in TagsInput
• Adjust prop types and imports

OlegIvaniv committed:
Remove console.logs

OlegIvaniv committed:
Enhance workflow evaluation UI with localization and styling improvements

• Add localization for UI text
• Improve styling and layout consistency
• Refactor component props and structure
• Update button labels and placeholders

OlegIvaniv committed:
Improve workflow evaluation UI and add localization for test list

- Update MainHeader tab logic
- Enhance EmptyState component styling
- Add localization for test list texts
- Adjust TestsList component
- Update locale file

OlegIvaniv committed:
Refactor evaluation form and improve error handling in workflow evaluation

• Remove dependency on tagsStore
• Add field-specific error handling
• Simplify loadTestData and saveTest methods
• Update EvaluationEditView component
• Improve UI feedback for form errors

OlegIvaniv committed:
Update evaluation UI components and add test IDs

- Add data-test-id to input components
- Update TagsInput to use useI18n composable
- Fix typo in TestsList component
- Add test ID to "Run Test" button
- Remove unused locale references

OlegIvaniv committed:
Add tests for WorkflowEvaluation components and EvaluationEditView

- Add MetricsInput component tests
- Add TagsInput component tests
- Add EvaluationEditView tests
- Include snapshot test
- Update test utilities and mocks

OlegIvaniv committed:
Remove unused CHAT_TRIGGER_NODE_TYPE import from viewsData.ts

OlegIvaniv committed:
Fix stray character

OlegIvaniv committed:
Remove unused view file

OlegIvaniv committed:
Enhance evaluation form with auto-save and improved error handling

- Add auto-save functionality
- Improve error handling in saveTest
- Update URL after creating new test

OlegIvaniv committed:
Enhance test definition creation and editing with description support

• Add description to test definition interface
• Include description in create/update params
• Update UI to handle description field
• Refactor types for better consistency
• Add data-test-id for E2E testing

OlegIvaniv committed:
Refactor evaluations API and add store tests

- Simplify API interface
- Update test definitions endpoint
- Add evaluations store unit tests
- Enhance upsert logic with validation
- Improve fetchAll method handling

OlegIvaniv committed:
Add useEvaluationForm tests and update evaluations store

- Added unit tests for useEvaluationForm
- Updated STORES enum to include EVALUATIONS
- Changed evaluations store id to use STORES.EVALUATIONS
- Updated testDefinitionsById type to Partial<ITestDefinition>
- Added deleteById method documentation

OlegIvaniv committed:
Remove TagsInput.test.ts

OlegIvaniv committed:
Revert rebased changed

OlegIvaniv committed:
Refactor workflow evaluation components to use useI18n composable

- Replace $locale with useI18n()
- Update component imports
- Fix type issues in useEvaluationForm
- Improve type safety in EvaluationListView
- Remove unused addFieldIssue function

OlegIvaniv committed:
Refactor workflow evaluation UI and introduce EvaluationStep component

- Add EvaluationStep component
- Reorganize layout and styling
- Remove redundant formGroup classes
- Adjust container width and structure
- Update content display and spacing

OlegIvaniv committed:
Refactor and enhance workflow evaluation UI components

- Redesign EvaluationStep component
- Add BlockArrow component for visual flow
- Update WorkflowSelector and MetricsInput
- Improve layout and styling of edit view
- Add new icons and update translations

OlegIvaniv committed:
Minor changes

burivuhster commented on packages/editor-ui/src/api/evaluations.ee.ts:
```diff
@@ -0,0 +1,64 @@
+
+// Query options type
+export interface ITestDefinitionsQueryOptions {
+	includeScopes?: boolean;
```
I think API does not support this yet

OlegIvaniv replied:
I got these from the CLI interface: https://github.com/n8n-io/n8n/blob/e875bc55925bd5e0bac3a8930f936bbea592dd0b/packages/cli/src/evaluation/test-definitions.types.ee.ts#L16

But will remove it for now

burivuhster commented on packages/editor-ui/src/api/evaluations.ee.ts:
```diff
@@ -0,0 +1,64 @@
+}
+
+export interface ITestDefinition extends ITestDefinitionBase {
+	id: number;
```
`string` now (#11742)

burivuhster commented on packages/editor-ui/src/api/evaluations.ee.ts:
```diff
@@ -0,0 +1,64 @@
+	return await makeRestApiRequest<ITestDefinition>(context, 'PATCH', `${endpoint}/${id}`, params);
+}
+
+export async function deleteTestDefinition(context: IRestApiContext, id: number) {
```
`string` now (#11742)

burivuhster commented on packages/editor-ui/src/views/WorkflowEvaluation/EvaluationEditView.vue:
```diff
@@ -0,0 +1,243 @@
+			<div :class="$style.panelIntro">When running a test</div>
+			<BlockArrow :class="$style.introArrow" />
+			<div :class="$style.panelBlock">
+				<EvaluationStep :class="$style.step" :title="'Fetch 5 past executions'">
```
Would it make sense to remove number of executions for now? Because fetching the real number might require some time, that might be spent on something more important

OlegIvaniv replied:
Will adjust this in the next stage — mocking nodes and selecting executions

burivuhster commented on packages/editor-ui/src/components/WorkflowEvaluation/ListEvaluation/TestsList.vue:
General suggestion: can we name these entities `Test Definition` for consistency? Not just here, but everywhere in the codebase.

burivuhster commented on packages/editor-ui/src/stores/evaluations.store.ee.ts:
```diff
@@ -0,0 +1,184 @@
+				},
+				{},
+			);
+			fetchedAll.value = true;
```
Why do we consider all test definitions fetched here?

OlegIvaniv replied:
Moved it to the fetching method instead

burivuhster commented on packages/editor-ui/src/stores/evaluations.store.ee.ts:
```diff
@@ -0,0 +1,184 @@
+				const defId = toUpsertDef.id;
+				if (!defId) throw Error('ID is required for upserting');
+				const currentDef = testDefinitionsById.value[defId];
+				if (currentDef) {
+					testDefinitionsById.value = {
+						...testDefinitionsById.value,
+						[defId]: {
+							...currentDef,
+							...toUpsertDef,
+						},
+					};
+				} else {
+					testDefinitionsById.value = {
+						...testDefinitionsById.value,
+						[defId]: toUpsertDef,
+					};
+				}
```
It's a matter of preference, but undefined in spread syntax doesn't do anything, so the code may be simplified to
```suggestion
					testDefinitionsById.value = {
						...testDefinitionsById.value,
						[defId]: {
							...currentDef,
							...toUpsertDef,
						},
					};
```

OlegIvaniv replied:
Good point! Thanks!

burivuhster commented on packages/editor-ui/src/components/WorkflowEvaluation/EditEvaluation/TagsInput.vue:
```diff
@@ -0,0 +1,98 @@
+					transparent
+				/>
+			</div>
+			<TagsDropdown
```
Can we somehow adapt this TagsDropdown for a single tag select use case? It opens list of tags when you click on the input with selected tag, making a false impression that this is a multi-tag input.

OlegIvaniv replied:
I'll do this in the next step — selecting executions and mocking node

burivuhster commented on packages/editor-ui/src/components/WorkflowEvaluation/ListEvaluation/TestItem.vue:
```diff
@@ -0,0 +1,137 @@
+		</div>
+
+		<div :class="$style.actions">
+			<n8n-tooltip v-for="action in actions" :key="action.icon" placement="top">
```
Is it possible to add a tiny delay before tooltip appears?

OlegIvaniv replied:
Sure, added 1s delay

burivuhster reviewed (changes requested):
Nicely done! 🙌 Just a couple of minor comments.
Main suggestion is to rename entities so that it reflect their backend/DB counterparts.

OlegIvaniv committed:
Refactor evaluation functionality and improve test definition handling

- Update test definition interfaces
- Modify evaluation form logic
- Enhance store operations for evaluations
- Adjust API calls and response handling
- Improve UI components and styling

OlegIvaniv committed:
Refactor evaluation types and fix type-related issues

• Update TestDefinitionRecord types
• Adjust component prop and emit types
• Modify store to use string IDs
• Update test cases for new types
• Fix type errors in various components

OlegIvaniv committed:
Simplify test definitions API and remove unused includeScopes parameter

- Remove ITestDefinitionsQueryOptions
- Update getTestDefinitions function
- Adjust store to remove includeScopes
- Update tests to reflect changes
- Clean up unused imports and parameters

OlegIvaniv committed:
Merge branch 'master' into ai-426-frontend-test-viewedit-screen_rebased

OlegIvaniv committed:
Rename WorkflowEvaluation to TestDefinition

OlegIvaniv committed:
Enhance test definition UI with i18n and improve list sorting

• Add i18n support for test definition UI
• Improve test list sorting by update time
• Enhance EvaluationStep component
• Update TestItem and TestsList components
• Add new localization strings

OlegIvaniv commented on packages/editor-ui/src/components/WorkflowEvaluation/types.ts:
```diff
@@ -0,0 +1,13 @@
+export interface TestExecution {
```
✅

OlegIvaniv committed:
Add delay to tooltip display in TestItem component

OlegIvaniv committed:
Fix unit tests

burivuhster reviewed (dismissed):
Thanks for addressing my comments!
LGTM 🚀

OlegIvaniv committed:
Quiet, linter

janober commented:
Got released with [`n8n@1.70.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.70.0)
